cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsProcessor.java

```diff
@@ -57,6 +65,21 @@ void addMetric(CruiseControlMetric metric) {
       brokerLoad.recordMetric(metric);
       return brokerLoad;
     });
+
+    // Compute cached number of cores by broker id if they have not been cached already.
+    Node node = cluster.nodeById(brokerId);
+    _cachedNumCoresByBroker.computeIfAbsent(brokerId, bid -> {
+      BrokerCapacityInfo capacity = _brokerCapacityConfigResolver.capacityForBroker(getRackHandleNull(node), node.host(), bid);
+      // No mapping shall be recorded if capacity is estimated, but estimation is not allowed.
+      return (!_allowCpuCapacityEstimation && capacity.isEstimated()) ? null : capacity.numCpuCores();
```

Comment:
I am thinking should we throw exception when we detect this condition since core information is a necessary thing to calculate cpu utilization. 

Anther thing to consider is should we try to cache core number in this method. For one thing, `AddMetric()` is call frequently; for another thing, the purpose of this method is to process reporter's metrics record. I think we can collect core information in `process()` method.

we can revisit this issue in the next patch.

Reply:
> I am thinking should we throw exception when we detect this condition since core information is a necessary thing to calculate cpu utilization.

I believe this information is only relevant when building a `PartitionMetricSample`, which will not be created if (1) there are no partitions in the cluster, or (2) sampling mode is `MetricSampler.SamplingMode.BROKER_METRICS_ONLY`. Hence, this part will be handled when creating partition samples (i.e. also added a `TODO` item for the next patch).

-- addressed the second feedback.